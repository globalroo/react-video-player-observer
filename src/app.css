.App {
	text-align: center;
}

.App-logo {
	animation: App-logo-spin infinite 20s linear;
	height: 60px;
}

.App-header {
	background-color: #222;
	padding: 10px;
	color: white;
}

.App-intro {
	font-size: medium;
}

@keyframes App-logo-spin {
	from {
		transform: rotate(0deg);
	}
	to {
		transform: rotate(360deg);
	}
}

.btn-group-max-width {
	width: 100%;
}

.now-playing {
	color: #00bc8c;
	text-transform: uppercase;
	font-weight: bold;
}

.player-observer-container {
	background-color: #333333;
	float: left;
	padding: 10px;
	padding-top: 20px;
	border: 1.5px solid #00bc8c;
	border-radius: 10px;
	min-width: 100%;
	box-shadow: 0px 0px 50px rgba(0,0,0,.6);
}

.panel {
	box-shadow: 0px 0px 50px rgba(0,0,0,.6);
}
