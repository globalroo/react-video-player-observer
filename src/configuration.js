export const HISTORICAL_EVENTS = 12;

export const getAvailableMedia = () => ([
	{
		name: "Sintel",
		poster: "http://media.w3.org/2010/05/sintel/poster.png",
		src: "https://media.w3.org/2010/05/sintel/trailer.mp4",
	},
	{
		name: "Big Buck Bunny - Trailer",
		poster: "http://media.w3.org/2010/05/bunny/poster.png",
		src: "http://media.w3.org/2010/05/bunny/trailer.mp4",
	},
	{
		name: "Big Buck Bunny - Movie",
		poster: "http://media.w3.org/2010/05/bunny/poster.png",
		src: "http://media.w3.org/2010/05/bunny/movie.mp4",
	},
]);

export const PLAYER_EVENTS_TO_OBSERVE = {
	ABORT: "abort",
	CANPLAY: "canplay",
	CANPLAYTHROUGH: "canplaythrough",
	DURATIONCHANGE: "durationchange",
	EMPTIED: "emptied",
	ENDED: "ended",
	ERROR: "error",
	LOADEDDATA: "loadeddata",
	LOADEDMETADATA: "loadedmetadata",
	LOADSTART: "loadstart",
	PAUSE: "pause",
	PLAY: "play",
	PLAYING: "playing",
	PROGRESS: "progress",
	RATECHANGE: "ratechange",
	RESIZE: "resize",
	SEEKED: "seeked",
	SEEKING: "seeking",
	STALLED: "stalled",
	SUSPEND: "suspend",
	TIMEUPDATE: "timeupdate",
	VOLUMECHANGE: "volumechange",
	WAITING: "waiting",
};
