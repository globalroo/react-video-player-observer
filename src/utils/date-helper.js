export const leftPad = (value, expectedLength = 2) => {
	let pad = value.length;
	let paddedValue = `${value}`;
	while (pad < expectedLength) {
		paddedValue = `0${paddedValue}`;
		pad += 1;
	}
	return paddedValue;
};

export const getFormattedTime = (date) => {
	const HH = leftPad(`${date.getHours()}`);
	const MM = leftPad(`${date.getMinutes()}`);
	const SS = leftPad(`${date.getSeconds()}`);
	const sss = leftPad(`${date.getMilliseconds()}`, 3);
	return `${HH}:${MM}:${SS}.${sss}`;
};
