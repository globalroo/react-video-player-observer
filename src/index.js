import React from "react";
import ReactDOM from "react-dom";
import "./index.css";
import { PlayerApp } from "./app";
import registerServiceWorker from "./registerServiceWorker";

import { getAvailableMedia } from "./configuration";

const availableMedia = getAvailableMedia();

ReactDOM.render(
	<PlayerApp
		availableMedia={availableMedia}
	/>,
	document.getElementById("root")
);
registerServiceWorker();
