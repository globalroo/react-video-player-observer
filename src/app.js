import "./app.css";

import React, { Component } from "react";
import PropTypes from "prop-types";

import { ControlsPanel } from "src/components/panels/controls-panel";
import { PLAYER_EVENTS_TO_OBSERVE, HISTORICAL_EVENTS } from "./configuration";
import { PlayerPanel } from "src/components/panels/player-panel";
import { StatsPanel } from "src/components/panels/stats-panel";
import { PlayerObserver } from "src/player-observer/player-observer";
import { getFormattedTime } from "src/utils/date-helper";

import logo from "./logo.svg";

const playerState = {
	currentTime: 0,
	remainingTime: 0,
	error: {
		code: 0,
		description: ""
	}
};

const selectedAsset = {
	name: "",
	poster: "",
	src: "",
};

export const initialState = {
	playerState,
	selectedAsset,
	eventHistory: []
};

export class PlayerApp extends Component {
	constructor(props) {
		super(props);
		this.state = initialState;
		this.eventListeners = [];
		this.observer = PlayerObserver.getInstance();
	}

	primePlayer(payload) {
		this._player.src = this.state.selectedAsset.src;
		this._player.poster = this.state.selectedAsset.poster;
	}

	onPlayerLoaded = (player) => {
		this._player = player;
		this.observer.initialisePlayer(player);
	}

	loadAsset = (selectedAsset) => {
		this.setState(
			() => ({ ...this.state, selectedAsset }),
			this.primePlayer
		);
	}

	addEventToHistory(event, oldEventHistory) {
		const eventHistory = [...oldEventHistory];
		const latestEvent = {
			timeStamp: getFormattedTime(new Date()),
			event,
		};
		if (eventHistory &&
			eventHistory.length === HISTORICAL_EVENTS) {
			eventHistory.pop();
		}
		eventHistory.unshift(latestEvent);
		return eventHistory;
	}

	handleState = (event, playerState) => {
		const eventHistory = this.addEventToHistory(event, this.state.eventHistory);
		const newState = { playerState, eventHistory };
		this.setState (newState);
	}

	subscribeEvents(trackEvents) {
		Object.keys(trackEvents).forEach((event) => {
			this.eventListeners.push(this.observer.subscribe(trackEvents[event], this.handleState));
		});
	}

	unsubscribeEvents() {
		this.eventListeners.forEach((eventUnsubscribeFn) => {
			eventUnsubscribeFn();
		});
	}

	componentDidMount() {
		this.subscribeEvents(PLAYER_EVENTS_TO_OBSERVE);
		this.loadAsset(this.props.availableMedia[0]);
	}

	componentWillUnmount() {
		this.unsubscribeEvents();
	}

	render() {
		return (
			<div className="App container">
				<div className="App-header col-xs-12">
					<img src={logo} className="App-logo" alt="logo" />
					<h4> React Player Observer Test Harness </h4>
				</div>
				<div className="player-observer-container">
					<div className="col-xs-12 col-sm-6">
						<PlayerPanel onPlayerLoaded={this.onPlayerLoaded} />
						<ControlsPanel asset={this.state.selectedAsset} loadAsset={this.loadAsset} availableMedia={this.props.availableMedia}/>
					</div>
					<div className="col-xs-12 col-sm-6">
						<StatsPanel playerState={this.state.playerState} eventHistory={this.state.eventHistory}/>
					</div>
				</div>
			</div>
		);
	}
}

PlayerApp.propTypes = {
	availableMedia: PropTypes.array,
};
