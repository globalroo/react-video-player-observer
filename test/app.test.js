import * as renderer from "react-test-renderer";
import React from "react";
import { PlayerApp, initialState } from "src/app";
import { getAvailableMedia, HISTORICAL_EVENTS } from "src/configuration";
import { mockDate, realDate } from "test/helpers/date-helper";

const testData = getAvailableMedia();

jest.mock("src/components/panels/controls-panel", () => ({
	ControlsPanel: () => <div> Controls Mock </div>
}));

jest.mock("src/components/panels/stats-panel", () => ({
	StatsPanel: () => <div> Stats Mock </div>
}));

jest.mock("src/components/panels/player-panel", () => ({
	PlayerPanel: ({ onPlayerLoaded }) => {
		onPlayerLoaded({
			src: "SRC",
			poster: "POSTER"
		});
		return <div> Player Mock </div>;
	}
}));

jest.mock("src/player-observer/player-observer", () => ({
	PlayerObserver: {
		getInstance: () => ({
			initialisePlayer: jest.fn(),
			subscribe: jest.fn(),
			unsubscribe: jest.fn()
		}),
	}
}));

describe("Application tests", () => {
	test("Application state is initialised as expected", () => {
		const component = new PlayerApp({ availableMedia: testData });
		expect(component.state).toBe(initialState);
	});

	test("Application state is updated by handleState (called from the observer) correctly", () => {
		const component = new PlayerApp({ availableMedia: testData });
		const newState = { "eventHistory": [{ "event": "test", "timeStamp": "14:00:00.000" }], "playerState": { "currentTime": 0 } };
		component.setState = jest.fn();

		mockDate(1506780000000); // Saturday, 30 September 2017 14:00:00
		component.handleState("test", { currentTime: 0 });
		expect(component.setState).toBeCalledWith(newState);
	});

	test("Application unsubscribes from the registered events in the observer", () => {
		const component = new PlayerApp({ availableMedia: testData });
		const unSubscribeFunction = jest.fn();
		component.eventListeners.push(unSubscribeFunction);
		component.unsubscribeEvents();
		expect(unSubscribeFunction).toHaveBeenCalledTimes(1);
	});

	test("Application cleans up after itself by removing event listeners when unmounting", () => {
		const component = new PlayerApp({ availableMedia: testData });
		component.unsubscribeEvents = jest.fn();
		component.componentWillUnmount();
		expect(component.unsubscribeEvents).toHaveBeenCalledTimes(1);
	});

	test("Application is rendered as expected", () => {
		const component = renderer.create(
			<PlayerApp availableMedia={testData} />
		);
		expect(component.toJSON()).toMatchSnapshot();
	});
});

describe("Event history works as intended", () => {
	afterEach(() => {
		(global).Date = realDate;
	});
	test("Add event history adds a new event", () => {
		const component = new PlayerApp({ availableMedia: testData });
		const expectedResult = [{ "event": "test", "timeStamp": "14:00:00.000" }];
		mockDate(1506780000000); // Saturday, 30 September 2017 14:00:00
		const newEventHistory = component.addEventToHistory("test", []);
		expect(newEventHistory).toEqual(expectedResult);
	});

	test("Add event history adds a new event to the front of the list and pops one off the end of the list", () => {
		const component = new PlayerApp({ availableMedia: testData });
		const fakeEvent = [{ "event": "test", "timeStamp": "13:59:00.000" }];
		const listSize = HISTORICAL_EVENTS;
		const fakeEvents = [];
		for (let i = 0; i < listSize; i++) {
			fakeEvents.push(fakeEvent);
		}
		mockDate(1506780000000); // Saturday, 30 September 2017 14:00:00
		const newEventHistory = component.addEventToHistory("test", fakeEvents);
		expect(newEventHistory).toMatchSnapshot();
		expect(newEventHistory[0].timeStamp).toBe("14:00:00.000");
		expect(newEventHistory).toHaveLength(HISTORICAL_EVENTS);
	});
});
