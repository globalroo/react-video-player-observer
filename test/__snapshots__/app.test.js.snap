// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Application tests Application is rendered as expected 1`] = `
<div
  className="App container"
>
  <div
    className="App-header col-xs-12"
  >
    <img
      alt="logo"
      className="App-logo"
      src="logo.svg"
    />
    <h4>
       React Player Observer Test Harness 
    </h4>
  </div>
  <div
    className="player-observer-container"
  >
    <div
      className="col-xs-12 col-sm-6"
    >
      <div>
         Player Mock 
      </div>
      <div>
         Controls Mock 
      </div>
    </div>
    <div
      className="col-xs-12 col-sm-6"
    >
      <div>
         Stats Mock 
      </div>
    </div>
  </div>
</div>
`;

exports[`Event history works as intended Add event history adds a new event to the front of the list and pops one off the end of the list 1`] = `
Array [
  Object {
    "event": "test",
    "timeStamp": "15:00:00.000",
  },
  Array [
    Object {
      "event": "test",
      "timeStamp": "14:59:00.000",
    },
  ],
  Array [
    Object {
      "event": "test",
      "timeStamp": "14:59:00.000",
    },
  ],
  Array [
    Object {
      "event": "test",
      "timeStamp": "14:59:00.000",
    },
  ],
  Array [
    Object {
      "event": "test",
      "timeStamp": "14:59:00.000",
    },
  ],
  Array [
    Object {
      "event": "test",
      "timeStamp": "14:59:00.000",
    },
  ],
  Array [
    Object {
      "event": "test",
      "timeStamp": "14:59:00.000",
    },
  ],
  Array [
    Object {
      "event": "test",
      "timeStamp": "14:59:00.000",
    },
  ],
  Array [
    Object {
      "event": "test",
      "timeStamp": "14:59:00.000",
    },
  ],
  Array [
    Object {
      "event": "test",
      "timeStamp": "14:59:00.000",
    },
  ],
  Array [
    Object {
      "event": "test",
      "timeStamp": "14:59:00.000",
    },
  ],
  Array [
    Object {
      "event": "test",
      "timeStamp": "14:59:00.000",
    },
  ],
]
`;
